packages/yarnpkg-fslib/sources/path.ts

```diff
@@ -81,30 +106,40 @@ export interface FormatInputPathObject<P extends Path> {
   name?: Filename;
 }
 
-export interface PathUtils<P extends Path> {
-  cwd(): P;
+export type MakeAbsolute<P> = P & (P extends PortablePath ? AbsolutePortablePath : AbsoluteNativePath);
+export type MakeRelative<P> = P & (P extends PortablePath ? RelativePortablePath : RelativeNativePath);
+
+export interface PathUtils<P extends Path, AP extends P = MakeAbsolute<P>, RP extends P = MakeRelative<P>> {
```

Comment:
Why introduce three generic parameters?
Great question!

In a previous iteration I used `MakeAbsolute<P>` instead of `AP` and `MakeRelative<P>` instead of `RP`, but that lead to various compilation errors.
This is also why `npath` and `ppath` are defined with explicit second and third generic parameter. The defaults are added to prevent this from being a breaking change and because in most cases it seems that explicit absolute and relative parameters are not needed.

An example of code that goes wrong:

```ts
declare function isValid(path: PortablePath): boolean;

function findPath(path: PortablePath | null) {
  if (path == null)
    return;

  while (path != PortablePath.root) {
    if (isValid(path)) {
      return path;
    }

    path = ppath.dirname(path);
  }
}
```

This will yield compilation errors saying `PortablePath | null` isn't assignable to `PortablePath`

Toy with this on the [Typescript playground](https://www.typescriptlang.org/play/#code/FAYw9gdgzgLgBAUwgVwLZwAoEMYAsAqAngA4JwDewccAYgJYA2CANFZmAE4xYBGTr1AHI46ANxbAAvsFCRYiFOmx4ASggYjRdGIQpsAgjyhgGyGBOpqNMMROnAEAD2Kd4O0nEPHT5jK95MyrhwALxwsBx0EADmcABketQA+sQ4uEnuCABcmGlEpAB09ExwAD65ePkIBX5cAQgA3GwpaUkc6prahDlBVp06BV4mZo1STU4uXHCZnkbD5sI24kGh4TCRMfGJcC14GSQIAPw9eQdFjGTlQVUFi7ZNyal77dZiXSeqHUtdg3M+o5Jxs5XNMDnA+ksELVuHwoWlVhEorEEpRHq1Mh8CGdipcKljCtD6g8dk90i9+t08RC3gNqeImoCHMCpjM6Qg7st4WFEZsUc1SftSMc8TccWURWcOaM0c8vjTKb05VpaUqAU0mZM3GDCbCVmEhv8dYF4eU2Ua4XggZrQR4pXrZt4RnaTeDVc68OKedF1RMQTMDSN7QGFpoLcFysGof5dWkrX6wWaua7XuJ3eHk6HzUF1TN7VmXWmfcytR4ALJYADWCEjAB4MAA+VYYLYACmbTnMEAAJlB2HUYx7Dg75lH+8aPTlI2mAJRxllg8tVtl1xthZsJNuIRydnt9mHj4JDs3Rg9wHJsmfqqLmDgAMywIDIQQAqjYGFA61ud72gsxPO3tyQXdmzCRdqz+EYV0bVFqDgdpvHEFsCmQ1IuCgScOA4LBCBXadJwwYlqC7OgOAgLBUAQGt8C-ICf3rFtSRyfA8LgfAGRkAB6DjWNwOhey7MAECgCAAHJ4AAd04CssmALsEBADR2jgcBoHgCBGLxV9GA-NN6yaLieL4uABKEuBJI4aTgAMuSFKwJSVPkdS0kxLT3xrNM-ynTM0j0qzuIQcQIGmXAwGQaJgh4MAPUyXs7LIPAyCcB94CgciWGmDhdBC8QNliMAcuCsgnI9UQ7LoeppjAcIEHi3B4rOGQbMUshb2QCAQBsSBwjACiAGFIGIzqIAYnIZxySKTAQLAIHVJq4rgJh4GIUbvI9coUAYBgr1vOAGLgABCMINoYadtmoAyqigEBImIeAQGmsS4CiW8EA4YKcDgYgnuEx7jrgabCFQTgEDYahxN4koW2MPqBu0OhIAY6dTpg2DYIM-Quy7f6Dv+tw6oULGwB2vAjIYKIyGmrGoGQTGkAYXRtDMqTe3aaI7K7JgoF7InCptMgeeK3BQdRz7VkFgp4JMRDxeI0i0sR2dhekSQgA)

Reply:
Just to be sure I understand - the TS playground you shared use the generic parameters, but still has the bogus behaviour you describe, is that expected? Or are those two different problems?

If this bug would affect us even with the generics, perhaps we should open an issue on the TS repo to see if they have an idea what might cause this.

Reply:
The playground link is a bit unclear, you have to scroll all the way down to see a comment:

```ts
// This doesn't work:
declare const npath: PathUtils<NativePath>;
// This does work:
// declare const npath: PathUtils<NativePath, AbsoluteNativePath>;
// even though both types are the exact same
```

if you comment the first declaration of `npath` and uncomment the second, the compilation error goes away even though the type of `npath` doesn't change.
Even without the bug I kinda prefer the version with three generic parameters because `AbsoluteNativePath` is easier to understand than `MakeAbsolute<NativePath>`